### PR TITLE
Accept Icon Type as a prop

### PR DESCRIPTION
**Pull request by yeknava**
In some situations I need to have different icon type for specific icons.

### CODE REVIEW

**Comment by alexrussell** [comment]
This would be super useful as it means we can use icons from other sets without having to manually import from `vector-icons` and lose the NB theme support.

**Comment by rabin111** [comment]
I need this! much more easier...

**Comment by SupriyaKalghatgi** [comment]
Related issues: #389

File: src/basic/IconNB.js [hunk 1]
```diff
@@ -65,6 +66,7 @@ class IconNB extends Component {
 
 IconNB.propTypes = {
 	style: PropTypes.oneOfType([PropTypes.object, PropTypes.number, PropTypes.array]),
+	type: PropTypes.string.oneOf(["Ionicons", "Entypo", "FontAwesome", "Foundation", "MaterialIcons", "MaterialCommunityIcons", "Octicons", "Zocial", "SimpleLineIcons"]),
```
Comment (alexrussell) [inline src/basic/IconNB.js]: Does `PropTypes.string.oneOf([])` work? I'd have though `PropTypes.oneOf([])`

**Comment by yeknava** [comment]
@alexrussell thank you, i fixed that. i thought all rules can be chain but it looks like its specific to isRequired.

**Comment by alexrussell** [comment]
np, just hoping we can get this merged :)

**Comment by SupriyaKalghatgi** [comment]
@yeknava Can you share example snippet with output for this

**Comment by alexrussell** [comment]
Sample intended use:

```jsx
<View>
    <Icon name='person' />
    <Icon type='FontAwesome' name='user' />
</View

//  Will actually 'translate' to the followoing:

<View>
    <Ionicons name='person' />
    <FontAwesome name='user' />
</View>

// (Assuming the icon theme is the default of Ionicons.)
```

The only two issues I see right now is that the entirety of `props` is proxied to the *actual* icon that'll be used, so actually my example above isn't quite correct, as the second example (Font Awesome) will also get the `type` prop passed along. Arguably, this should be removed from the props being proxied.

The second issue is that the `type` prop will only be used if the Icon has a theme set in the context. Now, I think a theme is passed in by default anyway so it's unlikely that there won't be one, but arguably the `type` prop should always set the icon type regardless of the context theme.

**Comment by alexrussell** [comment]
Also, setting `this.Icon` in `cWM` might be asking for trouble - if the `Icon` received a different type in props, the new type will not be used as the previous `Icon` instance won't be unmounted and remounted, but just re-rendered with the new props sent through `cWRP`.

Of course this isn't an issue with this PR per se, it's just the way that this component was originally set up (setting a class property in `cWM`), but this PR makes it meaningful that this is done as a class property in `cWM` so it is something that should be addressed by this PR.

That said, I'd say that for now the PR is fine to be merged, and the things I've mentioned can be documented as current caveats, which can be fixed later.

**Comment by yeknava** [comment]
@alexrussell i like your approach but i see this problem in any NB components. they lacks of customization when i compare them with [RNE](https://react-native-training.github.io/react-native-elements/) or other react or vuejs components. there are a lot of options as props and accept style also for more customization.

**Comment by alexrussell** [comment]
@yeknava sorry I don't quite understand - isn't this PR specifically to add customisation to the `Icon` component?

All I was saying above, for clarity, is that there's a potential issue with the `type` prop not being respected if it were to change dynamically during runtime. Consider the following code:

```jsx
render () {
  return (
    <View>
      <Icon type={this.state.iconTypeFA ? 'FontAwesome' : 'Ionicons'} name='arrow-up' />
      <Button onPress={() => this.setState({ iconTypeFA: ! Boolean(this.state.iconTypeFA) })}>
        Change icon type
      </Button>
    </View>
  )
}
```

That will render an "arrow-up" icon in Ionicons and a button underneath which switches the icon's type to 'FontAwesome'. But using the current code as in this PR, the icon's type will not update to FontAwesome as the `Icon` itself is not unmounted and remounted, but instead just sent new props and asked to re-render.

**Comment by yeknava** [comment]
@alexrussell nice catch, i addressed dynamic type change issue in patch-2 (didnt tested it yet btw)

**Comment by alexrussell** [comment]
Looks like it'd probably do the job, though I'd be inclined to actually go back to the original code that did the theme decision on mount and just do the props check in the `render` method - if `props.type` exists use that, otherwise use the `this.Icon` value we worked out before.

**Comment by yeknava** [comment]
@alexrussell its still get icon type from theme but in some cases we need to overwrite theme icon type for a specific icon. (i had a case like this)

**Comment by alexrussell** [comment]
Not sure what you're getting at there. No worries anyway, it's all good.

**Comment by OogieBoogieInJSON** [comment]
@alexrussell Some set of icons might not have the icon you need but the other one has it so we should be able to decide which icon set to use at component level.

**Comment by alexrussell** [comment]
@OogieBoogieInJSON I totally get that (I even gave the example of use that @SupriyaKalghatgi asked for), but that's what this PR does. I'm all for that.

**Comment by webdevsyd** [comment]
@alexrussell is this already implemented, i'm currently using native base 2.3.3

**Comment by SupriyaKalghatgi** [comment]
@webdevsyd this is not merged yet, that's why this PR is still open

**Comment by SupriyaKalghatgi** [comment]
@yeknava Please check
simulator screen shot - iphone 7 - 2018-01-12 at 15 19 40

**Comment by alexrussell** [comment]
@SupriyaKalghatgi @yeknava yeah looks like line 61 needs to be `this.setIcon` not just `setIcon`.

**Comment by yeknava** [comment]
@alexrussell thanks, it happens when changing code on github's text editor :smile:

**Comment by JeremyBradshaw7** [comment]
Works but getting this error in typescript:
image

**Comment by SupriyaKalghatgi** [comment]
@JeremyBradshaw7 #1594

**Comment by JeremyBradshaw7** [comment]
Thanks @SupriyaKalghatgi - this will hopefully be merged soon.

**Pull request merged**